src/rust/engine/fs/src/glob_matching.rs

```diff
@@ -157,246 +125,171 @@ trait GlobMatchingImplementation<E: Display + Send + Sync + 'static>: VFS<E> {
       return future::ok(vec![]).to_boxed();
     }
 
-    let init = PathGlobsExpansion {
-      context: self.clone(),
-      todo: include
-        .iter()
-        .flat_map(|entry| entry.to_sourced_globs())
-        .collect(),
-      exclude,
-      completed: IndexMap::default(),
-      outputs: IndexSet::default(),
-    };
-    future::loop_fn(init, |mut expansion| {
-      // Request the expansion of all outstanding PathGlobs as a batch.
-      let round = future::join_all({
-        let exclude = &expansion.exclude;
-        let context = &expansion.context;
-        expansion
-          .todo
-          .drain(..)
-          .map(|sourced_glob| context.expand_single(sourced_glob, exclude))
-          .collect::<Vec<_>>()
-      });
-      round.map(move |single_expansion_results| {
-        // Collect distinct new PathStats and PathGlobs
-        for exp in single_expansion_results {
-          let SingleExpansionResult {
-            sourced_glob: GlobWithSource { path_glob, source },
-            path_stats,
-            globs,
-          } = exp;
+    let result = Arc::new(Mutex::new(IndexSet::default()));
 
-          expansion.outputs.extend(path_stats.clone());
+    let mut sources = Vec::new();
+    let mut roots = Vec::new();
```

Comment:
`.zip()` in general makes me more uncomfortable than it should, but one way to rewrite this would be to use a vec of tuples and avoid the zipping below:
```rust
    let mut sources_with_roots = Vec::new();
    for pgie in include {
      let source = Arc::new(pgie.input);
      for path_glob in pgie.globs {
        sources_with_roots.push((source.clone(), 
                                 self.expand_single(result.clone(), exclude.clone(), path_glob)));
      }
      // ...
        let matching_inputs = sources_with_roots.iter()
        // ...
    }
```

This feels more clear to me, since the sources and roots are coming from the same place (and it feels easier to make edits to). It's not necessary at all.

Reply:
I thought about doing this initially, but I think that the issue was that the futures in `sources_with_roots` would need to move into the `join_all`... and that would mean unzipping it anyway.